[FLINK-21667][runtime] Defer starting ResourceManager to after obtaining leadership.

xintongsong opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*


## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*


## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

xintongsong commented:
@flinkbot run azure

wangyang0918 commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerServiceImpl.java:
```diff
@@ -0,0 +1,347 @@
+            ResourceManager<?> resourceManager) {
+        if (isLeader(resourceManager)) {
+            resourceManager.start();
+            forwardTerminationFuture(resourceManager);
```
I do not find it is problem, but it is a little confusing to me that completing `ResourceManager` will also complete the `serviceTerminationFuture`, which may cause the JobManager process to exit.

xintongsong replied:
IIUC, your point is that in case the leader RM is terminated unexpectedly, the RMService can try start another RM without failing the entire process?

In fact, it's not clear to me why would a leader RM be terminated unexpectedly without losing leadership. Ideally, this should not happen. Failing strictly rather than restarting silently in such cases may help drawing attentions to such unexpected cases.

Besides, failing strictly does not introduce any regression comparing to previous behaviors. Trying to start a new RM can be considered as an future improvement if it is proven necessary.

wangyang0918 reviewed (changes requested):
@xintongsong Thanks for creating this PR. I think it looks really good to me.

I have gone though the implementation carefully and left few comments. Except for these, I have one more concern about the `YarnResourceManagerDriver#initializeInternal`. After this change, the `ResourceManager` could be started more than once. We may get a fatal error when registering an existing application to Yarn.

Moreover, I will try to build this PR and run in a real Yarn/K8s cluster to check the behavior.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerServiceImpl.java:
```diff
@@ -18,63 +18,101 @@
+        this.fatalErrorHandler = rmProcessContext.getFatalErrorHandler();
+        this.ioExecutor = rmProcessContext.getIoExecutor();
+
+        this.handleLeaderEventExecutor = Executors.newSingleThreadExecutor();
```
True, it should be properly shutdown.

Yes. `grantLeadership` and `revokeLeadership` are invoked from the leader election service thread. Starting/stopping RMs may include external interactions. I think it would be nice to not block the leader election service with such operations.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerServiceImpl.java:
```diff
@@ -143,25 +147,16 @@ public void start() throws Exception {
-                // resource manager is never started
-                terminationFuture.complete(null);
-            }
+            FutureUtils.forward(previousResourceManagerTerminationFuture, serviceTerminationFuture);
```
I don't think so. If a future is completed multiple times, only the first time will take effect. In this case, the operations are synchronized, thus the forwarding will not complete the future before stopping leader election service.

xintongsong commented:
Thanks for the review, @wangyang0918.

Concerning re-registering AM on Yarn, I think it is indeed a problem. It looks like Yarn differentiate AMs by their attempt ids, and does not allow register multiple times for each AM. On the other hand, we need the registration response for finding existing containers. TBH, I don't see a good way to make the RM support multiple leader sessions on Yarn.

On the other hand, I think the purpose of this PR is to avoid a RM making changes that only should be made by the leader before obtaining the leadership. Maybe it makes sense to scope supporting multiple leader sessions out of this PR. We can either completely remove changes related to supporting multiple leader sessions, or introduce a toggle for this feature that is by default disabled. I'm slightly leaning towards the latter, which makes it easier for us to trying this out in future efforts.

WDYT?

tillrohrmann commented:
So this means that we would have to do the AM registration at the Yarn ResourceManager outside of Flink's ResourceManager which might be restarted when it loses and regains the leadership, right?

xintongsong commented:
@tillrohrmann 

Not necessarily. AFAIK, the Yarn RM requires each AM to register only once, but it does not require using the same `AMRMClient(Async)`. That means, in a later leader session, we can instantiate a new `AMRMClient(Async)` to interact with the Yarn RM, as long as the AM has been registered in the first leader session. This is my understanding based on the Yarn interfaces and docs, and would need further verifications to be sure. If this is proved true, we may simply catch and ignore the registration exception if it's caused by duplication, or maintain whether this is the first leader session in RMService (which is responsible for leader election and starting the RM on obtaining leadership) and skip the registration if not.

Another challenge I can see now is to handle resource changes between two leader sessions. Currently, the new leader RM relies on `RegisterApplicationMasterResponse` to find out what containers have already been allocated. This is based on the assumption that existing containers must came from the previous attempts. With multiple leader sessions, inheriting containers from previous leader sessions becomes non-trivial, because `AMRMClient(Async)` does not provide interfaces for getting all currently allocated containers. Two potential solutions are:
- Leveraging `YarnClient#getContainers`. `YarnClient` is meant to be used on the client side rather than by an AM. Not sure if there's any traps using it in an AM. Hopefully not.
- Alternatively, we can maintain Yarn specific component outside Flink's RM (as you've mentioned). The component can be reused across multiple Flink RMs / leader sessions. This component will be responsible for registering the AM to Yarn RM, as well as receiving the resource events from Yarn RM, which will forwarded to leader RM.

Personally, I think skipping registration if not the first leader session and leveraging `YarnClient` sounds promising. If it ends up we have to maintain something Yarn specific across multiple Flink RM's, I'm leaning towards to not introducing such complexity and keeping it as is as each attempt has only one leader session.

And again, I'd suggest to scope out this issue from the current PR.

wangyang0918 commented:
@xintongsong Thanks for the thorough analysis and solution. It makes sense to me to scope out the multiple leader sessions related code changes. After this PR, we could guarantee that only the active JobManager could start/stop the TaskManager pods/containers, as well as deregistering the application from cluster management. This is what we want to achieve originally in FLINK-21667.

To support multiple leader sessions("grant leadership" -> "revoke" -> "grant again") and make the JobManager process not crashed, I am not sure whether "Yarn duplicated register" is the only problem. We could have a follow-up ticket for more investigation.

> How to handle resource changes between two leader sessions?
I lean to the second solution. Using `YarnClient#getContainers` to get the full running containers and `AMRMClientAsync` heartbeat to get the incremental changes. It just feels like the K8s "list and watch" mechanism.

tillrohrmann commented:
What will currently happen if a RM loses the leadership? Will it simply terminate the process and let Yarn start a new process to reapply for the leadership?

I am fine with not letting the scope of this PR explode and rather split it up into several steps (some of which we might never implement maybe).

xintongsong commented:
> What will currently happen if a RM loses the leadership? Will it simply terminate the process and let Yarn start a new process to reapply for the leadership?

Yes, exactly.

xintongsong commented:
@wangyang0918,
I've addressed your comments, and rebased for conflict resolving. Please take another look.
I've also opened FLINK-22816 for investigating the possibility of multi-leader-session.

tillrohrmann commented:
> > What will currently happen if a RM loses the leadership? Will it simply terminate the process and let Yarn start a new process to reapply for the leadership?
> 
> Yes, exactly.

Sounds good to me.

wangyang0918 commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerServiceImpl.java:
```diff
@@ -208,6 +216,10 @@ public void revokeLeadership() {
                                 leaderSessionID);
 
                         stopLeaderResourceManager();
+
```
Here we will trigger a `FlinkException("Unexpected termination of ResourceManagerService"` in `DispatcherResourceManagerComponent` and then force the `ClusterEntrypoint` exited. After then, Yarn and K8s will launch a new container/pod for the JobManager.

wangyang0918 reviewed (approved):
Thanks for updating this PR.
LGTM +1 for merging.

I have also verified the failover behavior(lost leadership -> new JM process) and check the JobManager without suspicious logs.

tillrohrmann commented:
Unfortunately, I don't have time for a proper review. But I am confident that you'll manage to do it w/o me.
